connmgr/banmanager.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package connmgr
+
+import (
+	"net"
+	"sync"
+
+	"github.com/decred/dcrd/peer/v3"
+)
+
+// BanManagerConfig is the configuration struct for the ban manager.
+type BanManagerConfig struct {
+	// DisableBanning represents the status of disabling banning of
+	// misbehaving peers.
+	DisableBanning bool
+
+	// BanThreshold represents the maximum allowed ban score before
+	// misbehaving peers are disconnecting and banned.
+	BanThreshold uint32
+
+	// BanPeer represents a banning callback for cleanup and status
+	// updates purposes on a peer ban.
+	BanPeer func(p *peer.Peer)
+
+	// MaxPeers indicates the maximum number of inbound and outbound
+	// peers allowed.
+	MaxPeers int
+
+	// Whitelist represents the whitelisted IPs of the server.
+	WhiteList []net.IPNet
+}
+
+// banMgrPeer extends a peer to maintain additional state maintained by the
+// ban manager.
+type banMgrPeer struct {
+	*peer.Peer
+
+	isWhitelisted bool
+}
+
+// BanManager represents a peer ban score tracking manager.
+type BanManager struct {
+	cfg       BanManagerConfig
+	peers     map[string]*banMgrPeer
+	banScores map[string]*DynamicBanScore
+	mtx       sync.RWMutex
+}
+
+// NewBanManager initializes a new peer banning manager.
+func NewBanManager(cfg *BanManagerConfig) *BanManager {
+	return &BanManager{
+		cfg:       *cfg,
+		peers:     make(map[string]*banMgrPeer, cfg.MaxPeers),
+		banScores: make(map[string]*DynamicBanScore, cfg.MaxPeers),
+	}
+}
+
+// IsPeerWhitelisted checks if the provided peer is whitelisted.
+func (bm *BanManager) IsPeerWhitelisted(p *peer.Peer) bool {
+	host, _, err := net.SplitHostPort(p.Addr())
```

Comment:
This doesn't appear to be doing what we discussed.  I would expect this exported version to lookup the peer in the `peers` map (under the mutex of course) and return the field.

e.g. Assuming the other suggested changes:

```Go
func (bm *BanManager) IsPeerWhitelisted(p *peer.Peer) bool {
	bm.mtx.Lock()
	bmp := bm.lookupPeer(p)
	bm.mtx.Unlock()
	if bmp == nil {
		return false
	}
	return bmp.isWhitelisted
}
```

This code is obviously needed in order to originally set the field, but I would expect it to be unexported.